File: code/310最小高度树.cpp
/*
 * @lc app=leetcode.cn id=310 lang=cpp
 *
 * [310] 最小高度树
 */
#include "HF.h"
// @lc code=start
class Solution
{
public:
    vector<vector<int>> G; // 0 -> 1 2 3 4 5
    vector<int> res;
    bool vis[100005];
    int e[100005]; // 记录连通图的每个节点的离心率
    int bfs(vector<vector<int>> G, int s)
    {
        pair<int, int> ans = {s, 1};
        queue<pair<int, int>> Q;
        Q.push(ans);
        vis[s] = false;
        e[s] = max(e[s], 1);

        while (!Q.empty())
        {
            auto temp = Q.front();
            Q.pop();
            e[temp.first] = max(e[temp.first], temp.second);
            if (temp.second > ans.second)
            {
                ans = temp;
            }
            int u = temp.first;
            for (int i = 0; i < G[u].size(); i++)
            {
                int v = G[u][i];
                if (!vis[v])
                    continue;
                Q.push({v, temp.second + 1});
                vis[v] = false;
            }
        }
        return ans.first;
    }

    vector<int> findMinHeightTrees(int n, vector<vector<int>> &edges)
    {
        G.resize(n);
        for (auto &item : edges)
        {
            G[item[0]].push_back(item[1]);
            G[item[1]].push_back(item[0]);
        }
        memset(e, 0, sizeof(e));
        memset(vis, true, sizeof(vis));
        int end_p = bfs(G, 0);

        memset(vis, true, sizeof(vis));
        end_p = bfs(G, end_p);
        memset(vis, true, sizeof(vis));
        end_p = bfs(G, end_p);

        int min_e = INT32_MAX;
        for (int i = 0; i < n; i++)
        {
            min_e = min(min_e, e[i]);
        }
        for (int i = 0; i < n; i++)
        {
            if (e[i] == min_e)
                res.push_back(i);
        }

        return res;
    }
};
// @lc code=end

int main()
{

    return 0;
}

/*
6\n[[3,0],[3,1],[3,2],[3,4],[5,4]]


*/

File: code/310最小高度树.py
#
# @lc app=leetcode.cn id=310 lang=python3
#
# [310] 最小高度树
#
from utils.str2arr import *
# @lc code=start
class Solution:
    def findMinHeightTrees(self, n: int, edges):
        def bfs(G, s):
            ans = [s, 1]
            Q = []
            Q.append(ans)
            vis[s] = False
            e[s] = max(e[s], 1)

            while len(Q) > 0:
                temp = Q.pop(0)
                e[temp[0]] = max(e[temp[0]], temp[1])
                if temp[1] > ans[1]:
                    ans = temp[:]
                u = temp[0]

                for v in G[u]:
                    if not vis[v]: continue
                    Q.append([v, temp[1]+1])
                    vis[v] = False
            return ans[0]

        e = [0]*n
        vis = [True]*n
        G = [[] for _ in range(n)]
        for u, v in edges:
            G[u].append(v)
            G[v].append(u)

        end_p = bfs(G, 0)

        vis = [True]*n
        end_p = bfs(G, end_p)

        vis = [True]*n
        end_p = bfs(G, end_p)

        min_e = min(e)
        res = []
        for i, item in enumerate(e):
            if min_e == item:
                res.append(i)
        return res


# @lc code=end


n = int(input())
data = input()

edges = get_array(data)

test = Solution()
ans = test.findMinHeightTrees(n, edges=edges)

print(ans)